### PR TITLE
REST: verify existence of users by login or email

### PR DESCRIPTION
**Pull request by bdunogier**
> Implements [EZP-24820](https://jira.ez.no/browse/EZP-24820)
> Status: ready for review

Adds support for `login` and `email` query parameters on `GET /user/users`. Will return the user(s) matching the criteria. While login might only return one user, email might return several if non unique emails are allowed.

Also adds support for `HEAD` requests on the same resource. In that case, it will either respond with a `200` (yes, there is content matching the filter) or `404` (couldn't find anything). It doesn't differenciate between no permissions and no content.

Testing is done via behat, with a scenario for each filtering method. This required adding (primitive) Input Parsers for `UserList` and `UserRefList`.

## TODO
- [x] Update [specification](https://github.com/ezsystems/ezpublish-kernel/blob/master/doc/specifications/rest/REST-API-V2.rst#list-users)
- [x] Fix travis failure
- [x] ~~Create issue for the bugfix commit ("Fixed loading of user/group by roleId")~~

### CODE REVIEW

File: eZ/Bundle/EzPublishRestBundle/Features/User/load_users.feature [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    Scenario: Users can be filtered on login using a "login" query parameter
+       Given I have "administrator" permissions
+         And there is a user with the login "member"
+        When I create a "GET" request to "/user/users?login=member"
+         And I set header "Accept" with "UserList" object
+         And I send the request
+        Then response status code is 200
+         And response contains only the user with the login "member"
+
+    Scenario: Users can be filtered on login using a "login" query parameter
```
Comment (dpobel) [inline eZ/Bundle/EzPublishRestBundle/Features/User/load_users.feature]: this scenario is about the email it seems

File: eZ/Publish/Core/REST/Server/Controller/User.php [hunk 2]
```diff
@@ -457,11 +458,29 @@ public function loadUsers(Request $request)
...
+                    $this->userService->loadUser(
+                        $this->contentService->loadContentInfoByRemoteId($request->query->get('remoteId'))->id
+                    )
+                ),
+            );
+        } elseif ($request->query->has('login')) {
+            $restUsers = array(
+                $this->buildRestUserObject($this->userService->loadUserByLogin($request->query->get('login'))),
+            );
+        } elseif ($request->query->has('login')) {
```
Comment (pspanja) [inline eZ/Publish/Core/REST/Server/Controller/User.php]: This is basically repeats the previous condition.

Comment (bdunogier) [inline eZ/Publish/Core/REST/Server/Controller/User.php]: My, it does, does it not ? Thank you :)

**Comment by bdunogier** [comment]
HEAD is implemented.
Finished behat tests, including for the `HEAD` case, with and without permissions.

File: eZ/Publish/Core/REST/Server/Values/OK.php [hunk 3]
```diff
@@ -0,0 +1,12 @@
+<?php
+/**
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+namespace eZ\Publish\Core\REST\Server\Values;
+
+/**
+ * Value object for a 200 OK response.
+ */
+class OK
```
Comment (dpobel) [inline eZ/Publish/Core/REST/Server/Values/OK.php]: funny class but it should probably extend RestValue like NoContent

**Comment by bdunogier** [comment]
I can't figure out why the tests fail on Travis. They work just fine here, with a fresh install + ezpublish-kernel@this branch, with the api running in the behat env and a clean install via ezplatform:install.

Maybe somebody else could try ? Ping @miguelcleverti

**Comment by andrerom** [comment]
And you run the whole suite locally in the same way?

**Comment by miguelcleverti** [comment]
@bdunogier I recall something similar happening a while ago with the behat tests for content, I think it was related to cache @joaoinacio knows more about it, but I will try to run the tests.

**Comment by bdunogier** [comment]
> And you run the whole suite locally in the same way?

afaict yes:

- Fresh clone of ezplatform@master
- composer install 
- replaced kernel with my branch
- ezplatform:install clean

**Comment by bdunogier** [comment]
The last commit should fix the problem. I'm quoting the commit message as the explanation:

*Made GET /user/users 404 if empty*

If we don't do that, HEAD and GET won't behave the same way, which conflicts with the HTTP spec.

It is also inconsistent with Symfony's HTTP cache behaviour: upon a miss, the request is cloned, and forwarded as a GET one. The reasoning is that since `GET` and `HEAD` have the same behaviour, sending a GET will make sure we can cache both the body and headers.

**Comment by andrerom** [comment]
+1 besides need to squash commits and handle remaining tasks :)

**Comment by bdunogier** [comment]
About remaining task, there is still the "bugfix" mentioned in the description (56aef9c). I'm thinking as I'm writing it that it's not really a bugfix as it matches the specs.

The thing is about how ids are passed as GET parameters: `roleId=5` vs. `roleId=/user/roles/5`. The patch changes it to the numerical id version.

File: doc/specifications/rest/REST-API-V2.rst [hunk 4]
```diff
@@ -4888,7 +4890,28 @@ List Users
           User_
 
 :Error Codes:
-    :401: If the user has no permission to read users
+    :404: If there are no visibile users matching the filter
```
Comment (andrerom) [inline doc/specifications/rest/REST-API-V2.rst]: I'm a bit unsure about this change, for two reasons:
- 401 as I read it is a bout lack of permissions to read users at all, we could for instance check if user has rights to read any users at all early in the request, and if users are ever moved out of content tree it will have it's own module/policy making this more inline
- I don't think the 401 was intended to be thrown if no users where found, o I think 401 should still be in the spec, however given the nature of this endpoint I'm struggling a bit with agreeing with myself if no users should be 404 or just a empty list of user refs. As a developer I would expect 404 on unique id lookups, but empty lists on group lookups like roleId. However maybe I'm thinking to much PHP api here, and also given both cases are on the same endpoint I of course prefer it is consistent.


So to sum up:
- Keep 401 imho, as it from spec perspective seems to be about permissions, not NotFound cases. can be unimplemented atm.
- Add 404 as you did, or return empty list, I think you and @dpobel knows the needs of the ui better here and what works best in the REST api.

**Comment by andrerom** [comment]
> About remaining task, there is still the "bugfix" mentioned in the description (56aef9c). I'm thinking as I'm writing it that it's not really a bugfix as it matches the specs.
>
>The thing is about how ids are passed as GET parameters: roleId=5 vs. roleId=/user/roles/5. The patch changes it to the numerical id version.

Ok, cause spec says ` roleId=/user/roles/5` quite clearly, unsure why, but this is how id's are referred to often in the case of the REST clients (PHP and js), right?

**Comment by bdunogier** [comment]
> Ok, cause spec says roleId=/user/roles/5 quite clearly, unsure why, but this is how id's are referred to often in the case of the REST clients (PHP and js), right?

Yep, it is part of the discovery thing. But these don't take the prefix into account...
Anyway, I can move it away, but I need to move a couple other things then.

File: doc/specifications/rest/REST-API-V2.rst (hunk 4 above)
Comment (bdunogier) [inline doc/specifications/rest/REST-API-V2.rst]: > I'm a bit torn on if the new not found case should be 404 or just a empty user ref list, in case of roleId I would have expected the latter, while in case of unique ids I would have expected the former :P

I've been a bit torn as well. But overall, we don't have so many solutions: if we go for empty list, it means we send a 200, meaning there is no way to distinguish results with HEAD.

I concluded that if you ask for users matching a filter, and none matches, getting responded "not found" kinda makes sense. And it doesn't really break anything.

**Comment by bdunogier** [comment]
About the commits, I'll autosquash the last 2 ones, but all of the other ones are kind of atomic (they were amended overtime).

**Comment by bdunogier** [comment]
All squashed, and change to roleId filtering reverted, we should be fine.
